contracts/EthlanceJobs.sol

```diff
@@ -0,0 +1,851 @@
+  /// @param _sender the sender of the transaction issuing the job (should be the same as msg.sender unless the txn is called by the meta tx relayer)
+  /// @param _jobId the index of the job
+  /// @param _amount the amount of tokens being contributed
+  function contribute(
+                      address payable _sender,
+                      uint _jobId,
+                      uint _amount)
+    public
+    payable
+    senderIsValid(_sender)
+    validateJobArrayIndex(_jobId)
+    callNotStarted
+  {
+    require(_amount > 0); // Contributions of 0 tokens or token ID 0 should fail
+
+    jobs[_jobId].contributions.push(
+                                    Contribution(_sender, _amount, false)); // Adds the contribution to the job
+
+    if (jobs[_jobId].tokenVersion == 0){
+
+      jobs[_jobId].balance = jobs[_jobId].balance.add(_amount); // Increments the balance of the job
+
+      require(msg.value == _amount);
+    } else if (jobs[_jobId].tokenVersion == 20){
+
+      jobs[_jobId].balance = jobs[_jobId].balance.add(_amount); // Increments the balance of the job
+
+      require(msg.value == 0); // Ensures users don't accidentally send ETH alongside a token contribution, locking up funds
+      require(IERC20(jobs[_jobId].token).transferFrom(_sender,
+                                                      address(this),
+                                                      _amount));
+    } else if (jobs[_jobId].tokenVersion == 721){
+      tokenBalances[_jobId][_amount] = true; // Adds the 721 token to the balance of the job
+
+
+      require(msg.value == 0); // Ensures users don't accidentally send ETH alongside a token contribution, locking up funds
+      IERC721(jobs[_jobId].token).transferFrom(_sender,
+                                               address(this),
+                                               _amount);
+    } else {
+      revert();
+    }
+
+    emit ContributionAdded(_jobId,
+                           jobs[_jobId].contributions.length - 1, // The new contributionId
+                           _sender,
+                           _amount);
+  }
+
+  /// @dev refundContribution(): Allows users to refund the contributions they've
+  ///                            made to a particular job, but only if the job
+  ///                            has not yet paid out
+  /// @param _sender the sender of the transaction issuing the job (should be the same as msg.sender unless the txn is called by the meta tx relayer)
+  /// @param _jobId the index of the job
+  /// @param _contributionId the index of the contribution being refunded
+  function refundContribution(
+                              address _sender,
+                              uint _jobId,
+                              uint _contributionId)
+    public
+    senderIsValid(_sender)
+    validateJobArrayIndex(_jobId)
+    validateContributionArrayIndex(_jobId, _contributionId)
+    onlyContributor(_sender, _jobId, _contributionId)
+    hasNotPaid(_jobId)
```

Comment:
Is this needed in the context of jobs? Seems like jobs have recurring payments, so how should the refunding work for Jobs? Bounties allow only for a one-time refund